docs/content/concepts/configuration/advanced-config-types.mdx

```diff
@@ -0,0 +1,276 @@
+
+class MyAssetConfig(Config):
+    person_name: Optional[str] = None
+    greeting_phrase: str = "hello"
+
+@asset
+def greeting(config: MyAssetConfig) -> str:
+    if config.person_name:
+        return f"{config.greeting_phrase} {config.person_name}"
+    else:
+        return config.greeting_phrase
+
+asset_result = materialize(
+    [greeting],
+    run_config=RunConfig({"greeting": MyAssetConfig()}),
+)
+```
+
+---
+
+## Basic data structures
+
+Many basic Python data structures can be used in your config schemas, including lists and mappings.
+
+For example, we can define a config schema that takes in a list of user names and a mapping of user names to user scores.
+
+```python file=/guides/dagster/pythonic_config/pythonic_config.py startafter=start_basic_data_structures_config endbefore=end_basic_data_structures_config dedent=4
+from dagster import Config, materialize, asset, RunConfig
+from typing import List, Dict
+
+class MyDataStructuresConfig(Config):
+    user_names: List[str]
+    user_scores: Dict[str, int]
+
+@asset
+def scoreboard(config: MyDataStructuresConfig):
+    ...
+
+result = materialize(
+    [scoreboard],
+    run_config=RunConfig(
+        {
+            "scoreboard": MyDataStructuresConfig(
+                user_names=["Alice", "Bob"],
+                user_scores={"Alice": 10, "Bob": 20},
+            )
+        }
+    ),
+)
+```
+
+---
+
+## Nested schemas
+
+Schemas can be nested in one another, or in basic Python data structures.
+
+Here, we define a schema which contains a mapping of user names to complex user data objects.
+
+```python file=/guides/dagster/pythonic_config/pythonic_config.py startafter=start_nested_schema_config endbefore=end_nested_schema_config dedent=4
+from dagster import asset, materialize, Config, RunConfig
+from typing import Dict
+
+class UserData(Config):
+    age: int
+    email: str
+    profile_picture_url: str
+
+class MyNestedConfig(Config):
+    user_data: Dict[str, UserData]
+
+@asset
+def average_age(config: MyNestedConfig):
+    ...
+
+result = materialize(
+    [average_age],
+    run_config=RunConfig(
+        {
+            "average_age": MyNestedConfig(
+                user_data={
+                    "Alice": UserData(age=10, email="alice@gmail.com", profile_picture_url=...),
+                    "Bob": UserData(age=20, email="bob@gmail.com", profile_picture_url=...),
+                }
+            )
+        }
+    ),
+)
+```
+
+---
+
+## Union types
+
+Union types are supported using Pydantic [discriminated unions](https://docs.pydantic.dev/usage/types/#discriminated-unions-aka-tagged-unions). Each union type must be a subclass of <PyObject object="Config"/>. The `discriminator` argument to <PyObject object="Field"/> specifies the field that will be used to determine which union type to use.
+
+Here, we define a config schema which takes in a `pet` field, which can be either a `Cat` or a `Dog`, as indicated by the `pet_type` field.
```

Comment:
Might be worth detailing in-launchpad behavior/yaml/config dictionary behavior since it is non-obvious